### PR TITLE
fix: fix some remaining candid issues

### LINKED ISSUE #3519: FIX ME special opt rule hit!
Just hit this message when deploying to a canister. I was purposely changing the type signature of my API (breaking change), and hit this message instead of signaling a dfx deploy warning.

```
FIX ME! vec record { text; AttributeValue } <: opt UserMetadata/1 via special opt rule.
This means the sender and receiver type has diverged, and can cause data loss.
```

### PR DESCRIPTION
**Pull request by chenyan-dfinity**
# Description

* Support `import` for local did file. Fix SDK-746 and #2874 
* Subtyping check reports special opt rule as error, so that developers can confirm the type in the upgrade. Fix #3519 #3003 

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

# Checklist:

- [ ] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [ ] I have edited the CHANGELOG accordingly.
- [ ] I have made corresponding changes to the documentation.

### CODE REVIEW

File: src/dfx/src/lib/models/canister.rs [hunk 1]
```diff
@@ -338,30 +338,40 @@ fn wasm_opt_level_convert(opt_level: WasmOptLevel) -> OptLevel {
...
 
 fn separate_candid(path: &Path) -> DfxResult<(String, String)> {
+    use candid::pretty::candid::{compile, pp_args};
+    use candid::types::internal::TypeInner;
+    use candid_parser::{
+        pretty_parse,
+        types::{Dec, IDLProg},
+    };
+    let did = dfx_core::fs::read_to_string(path)?;
+    let prog = pretty_parse::<IDLProg>(path.to_str().unwrap(), &did)?;
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/models/canister.rs:348]: Let's avoid unwrap if we can.  Something like this, or return an error.
```suggestion
    let prog = pretty_parse::<IDLProg>(&format!("{}", path.display()), &did)?;
```

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]
Would you please add e2e tests that demonstrate what's being fixed here, and would detect regressions?

File: e2e/tests-dfx/metadata.bash [hunk 2]
```diff
@@ -166,9 +166,9 @@ teardown() {
   dfx canister metadata prebuilt_custom_blank_build candid:service >from_canister.txt
   diff main.did from_canister.txt
 
-  # this canister has a build step, but it is an empty array, so dfx leaves the candid:service metadata as-is
-  dfx canister metadata prebuilt_custom_empty_build candid:service >from_canister.txt
-  diff main.did from_canister.txt
```
Comment (ericswanson-dfinity) [inline e2e/tests-dfx/metadata.bash:171]: Is there a reason to delete this part of this test?  This test is verifying existing behavior.

Comment (chenyan-dfinity) [reply e2e/tests-dfx/metadata.bash]: Fixed.

**Review by ericswanson-dfinity — APPROVED** [review]

**Pull request merged**